### ISSUE TITLE
Multiple callbacks even if the location is within stationary radius

### ISSUE THREAD

**Issue by shyamal890** [mauron85/cordova-plugin-background-geolocation#90]
Following is the configuration I have set:

```

            locationProvider: backgroundGeoLocation.provider.ANDROID_DISTANCE_FILTER_PROVIDER,  //Only Android
            debug: true, // <-- enable this hear sounds for background-geolocation life-cycle.
            stopOnTerminate: false, // <-- enable this to clear background location settings when the app terminates
            startOnBoot: true,  //Android Only, starts location tracking on device boot

            desiredAccuracy: 100,    //In Meters
            stationaryRadius: 60,  //30,  //In Meters
            distanceFilter: 40, // 20,     //In Meters, minimum distance a user must move horizontally before update
            locationTimeout: 300,    //Android Only. Minimum time in secs before location update
```

However, even when the phone is stationary it receives random callbacks at random time. Isn't `stationaryRadius` of 60 be taken into account?

**Comment by mauron85** [comment]
What platform and version of the plugin? In 2.0.0 locationTimeout was renamed just to timeout.

**Comment by shyamal890** [comment]
You mean `interval` right?

Forgot to update to `interval`. Will inform, if I encounter the same result after update.

**Comment by mauron85** [comment]
Yes interval.

**Comment by shyamal890** [comment]
I can confirm that this issue persists even after the change. Moreover, `interval`too is not considered as callbacks are less than the `interval` defined.

**Comment by mauron85** [comment]
Do you think it is behaving differently than it version 1.0.x?

**Comment by shyamal890** [comment]
No I haven't tested on 1.0.x. I will get back with more results on 2.0.x

**Comment by shyamal890** [comment]
@mauron85 No I haven't tested with 1.0.x . 

But to be more specific this is an example of what is happening in stationary condition. I have set the stationary radius to 60 meters, yet I am receiving callbacks which are within 60 meters. 

![i](https://cloud.githubusercontent.com/assets/6606965/16205573/0ee586f6-3742-11e6-87ef-bd5b9121e0a5.PNG)

**Comment by mauron85** [comment]
Ok then, I need to check code for ANDROID_DISTANCE_FILTER_PROVIDER.

**Comment by shyamal890** [comment]
@mauron85 Please do look at this, this is the latest spanshot. It is so bad that I cannot even show this to anyone.

I wonder if Chris' plugin use geofencing for this reason?

![i1](https://cloud.githubusercontent.com/assets/6606965/16540642/7b446092-4089-11e6-947f-045283b76037.PNG)

**Comment by mauron85** [comment]
ANDROID_DISTANCE_FILTER is practically untouched Chris implementation for Android. It does uses geofences (locationManager.addProximityAlert).

**Comment by mauron85** [comment]
Can you post locations json here?

**Comment by shyamal890** [comment]
@mauron85 We can in code filter based on distance between 2 consecutive locations. However, it is something that plugin by itself should be doing isn't it? 

Anyways, I have these locations stored in database, would you like data in a csv file?

**Comment by mauron85** [comment]
The question here is if it's a bug or device specific behaviour. Honestly without deeper investigation (as I didn't wrote that provider) I don't know.

**Comment by shyamal890** [comment]
I tested it on three devices Moto G-3 Turbo, Xiaomi Mi4i and Samsung Galaxy Grand. All have the same issue, so this behavior is device independent.

**Comment by mauron85** [comment]
On your picture, what represents the circle? Is it accuracy?

**Comment by shyamal890** [comment]
Oh no no, change in icon color red and green is dependent on speed recorded. Nothing to do with accuracy.

**Comment by shyamal890** [comment]
Did you get the same result on your end?

**Comment by mauron85** [comment]
I did some investigation on your data. Here are the conclusions.

Comparison of speed from sensor and calculated speed: Speed was off limit at few moments, but don't think it's causing your issue.

![screenshot 2016-07-05 00 45 18](https://cloud.githubusercontent.com/assets/359450/16570656/4e82a6d2-424a-11e6-8bd6-6f6383cad4b9.png)

There are no extreme spikes in accuracy.

![screenshot 2016-07-05 00 51 12](https://cloud.githubusercontent.com/assets/359450/16570668/9821992e-424a-11e6-9a46-01987fbfc3ff.png)

So there is really only problem with rendering your route on a map. I believe ANDROID_DISTANCE_PROVIDER with its geofencing implementation is not for you. I would say, geofences are not accurate enough for your use case. Try use ANDROID_ACTIVITY_PROVIDER instead.

Also you try to smoothen paths with some kind of algorithm. There is nice thread on stackoverflow.
http://stackoverflow.com/questions/1134579/smooth-gps-data

**Comment by mauron85** [comment]
>if( SquareRoot( (X2 - X1)^2 + (Y2 - Y1)^2 ) > distanceFilter){
     //Then only record the new location
}

What is the theory behind this? Any link? Would like to learn more.

**Comment by shyamal890** [comment]
The formula finds distance between 2 points http://www.mathwarehouse.com/algebra/distance_formula/index.php

**NOTE**: I am only facing this multiple callbacks issue in stationary condition. 

So, this is how it would work:
When a new location update is available the plugin would calculate the distance between the last location and new one. If the distance is less than `stationaryRadius` ignore the new location. This would make sure that the plot like above don't happen. Plugin must only check distance in stationary condition( I am guessing there is a way to know if a device is stationary)

**Comment by mauron85** [comment]
I'm definitely thinking about filters like Kalman in the future. But question for now is, if there is a bug in ANDROID_DISTANCE_FILTER implementation or it's just the geofences which are not suitable for this kind of use cases. Anyway soon we will know the answer. I've implemented db logging of all events in iOS version (next-ios) branch and soon also Android logs will be implemented. Then we can observe what the hell is going on.

**Comment by shyamal890** [comment]
@mauron85 There is no this kind of usage :P This is something everyone may be facing when the device is stationary. 

So I humbly request you to seriously consider adding this distance formula check in stationary location callback.(Must be affecting every user of this plugin) May be we can have a flag activating this check.

**Comment by mauron85** [comment]
Your use case means that you want reconstruct **exact** user movement. You should consider using ANDROID_ACTIVITY_PROVIDER in such a case.  Appropriate use case for ANDROID_DISTANCE_FILTER is probably only when you don't have very high demands on accuracy of user location changes, but prefer battery savings over accuracy.

**Comment by shyamal890** [comment]
No, I am pretty happy with location accuracy being 100m (check the code in the first comment). That is why I have set stationaryRadius to 60m. The problem is that locations callbacks are within this 60m of radius in stationary condition! 

Location callbacks are alright when the device is in moving condition.

**Comment by mauron85** [comment]
I'm also talking about stationary mode. ANDROID_ACTIVITY_PROVIDER  doesn't have it so if you use, your problem is gone.

**Comment by shyamal890** [comment]
Ok, so with ANDROID_ACTIVITY_PROVIDER is `stopOnStillActivity` the feature that would avoid multiple callbacks in stationary condition?

**Comment by mauron85** [comment]
There is no stationary condition in ANDROID_ACTIVITY_PROVIDER. You normally get location updates same way as in foreground. `stopOnStillActivity` only turns off activity detection - way to detect if device is IDLE (not moving).

**Comment by shyamal890** [comment]
Since, you insisted on ANDROID_ACTIVITY_PROVIDER I did a trial. See the result in the pic below. It is the same as Distance provider, no difference!

Again, the problem is location updates shouldn't happen if the new location is within the stationary radius defined in distance provider (when the device is in stationary condition). This can be solved by implementing the distance formula I provided earlier.  

![i1](https://cloud.githubusercontent.com/assets/6606965/16616992/934689c8-439e-11e6-9a17-f14320f926a7.PNG)

**Comment by uzumakinaruto123** [comment]
I am using the same distance formula. But ,my use case is little different . I am here just to confirm that while using `ANDROID_ACTIVITY_PROVIDER` with `stopOnStillActivity:true` and `interval:30000` , will it always give me user location ` every 30 secs` irrespective of activity and `only once` ? @mauron85

**Comment by mauron85** [comment]
I'm repeating my self, but there is no stationary condition in ANDROID_ACTIVITY_PROVIDER.

**Comment by shyamal890** [comment]
So, how does Activity provider solve the problem we are faced with? You may be giving a solution but it is not addressing the problem. 

Sent from Yahoo Mail on Android 
 
  On Wed, 6 Jul, 2016 at 10:54 p.m., Marián Hello<notifications@github.com> wrote:   
I'm repeating mself , but there is no stationary condition in ANDROID_ACTIVITY_PROVIDER.

**Comment by mauron85** [comment]
ANDROID_ACTIVITY_PROVIDER is just giving you positions by time defined `option.interval` when you're moving and will turn off location service when you stop moving. Nothing more or less. How you process them, is up to you.

**Comment by shyamal890** [comment]
@mauron85 

> From principle it should be more accurate then the other provider, which is using geofences, but it doesn't mean it is silver bulet. You still has to process your results.

Since, there is a `stationaryRadius` configuration in Distance provider, it inherently says that while mobile is stationary no locations within `stationaryRadius` should be recorded. However, that is not the case infact quite opposite as per the data I sent earlier. 

> What are you suggesting with the formula is actually your app/server task and can be easily implemented on your side

I agree this can be done on our side. However, I see everyone facing the same issue, may be I am wrong and would leave it there.

**Comment by mauron85** [comment]
> However, that is not the case infact quite opposite as per the data I sent earlier.

**Comment by shyamal890** [comment]
@mauron85 Do check Christocracy's implementation it handles this issue for both the providers.

**Comment by mauron85** [comment]
@shyamal890 I don't have access to his private implementation.

**Comment by mauron85** [comment]
@shyamal890 Anyway if you're willing to help. Please try latest dev version with logging capability.

1. Install dev version from git
```
cordova plugin add https://github.com/mauron85/cordova-plugin-background-geolocation#next-android
```

2. Enable `option.debug`

3. Print logs to chrome console - https://github.com/mauron85/cordova-plugin-background-geolocation/tree/next-android#debugging). Or send them to your server and then post here copy.

```
function postLogs(logEntries) {
  function jQueryPost(logEntries) {
    logEntries = logEntries || [];
    console.log('Posting ' + logEntries.length + ' log entries...');
    jQuery.ajax({
      type: 'POST',
      url: 'http://192.168.81.15:3000/debug',
      data: JSON.stringify(logEntries),
      timeout: 5000
    })
    .done(function () {
      console.log('Logs has been successfully posted.');
    })
    .fail(function(xhr) {
      console.log('Posting of logs failed with status: ' + xhr.status);
    });
  }
  // will load jQuery if you not using it in your app
  if (!window.jQuery) {
    var script = document.createElement('script');    
    script.src = 'https://ajax.googleapis.com/ajax/libs/jquery/3.0.0/jquery.min.js';
    script.onload = function () { jQueryPost(logEntries); };
    document.getElementsByTagName('head')[0].appendChild(script);
  } else {
    jQueryPost(logEntries);
  }
}

backgroundGeolocation.getLogEntries(100, postLogs);
```

**Comment by shyamal890** [comment]
@mauron85 Can you switch off the debug sound or set a different config flag for logging? I would like to test this in real scenario, will give better data points to debug.

**Comment by mauron85** [comment]
@shyamal890 Do i understand you correctly? You would like to record debug logs also in 'production' mode?

**Comment by shyamal890** [comment]
Yes that's right

**Comment by mauron85** [comment]
Any updates on this? 2.2.x does not make sounds and always logs, so making logs should be possiable in prod mode.

**Comment by shyamal890** [comment]
@mauron85 I have stopped using Distance provider and am using Fused location provider now.

**Comment by mauron85** [comment]
This issue is almost 2 years old. Closing. Please reopen if needed.
